### PR TITLE
 [Config] Backport string|null api for node names

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes    
| Fixed tickets | #...   
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

**Review by Simperfit — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
For 2.7

**Comment by ro0NL** [comment]
Fixed `ParentNodeDefinitionInterface` api docs as well :)

**Review by Tobion — APPROVED** [review]

File: src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php [hunk 1]
```diff
@@ -143,8 +143,8 @@ public function end()
     /**
      * Creates a child node.
      *
-     * @param string $name The name of the node
-     * @param string $type The type of the node
+     * @param string|null $name The name of the node
```
Comment (Tobion) [inline src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php]: It does not make sense to change this when you don't change all the other places as well like https://github.com/ro0NL/symfony/blob/2d72b61613131160d2f4716f71bae7a55a9f4153/src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php#L39 and https://github.com/ro0NL/symfony/blob/2d72b61613131160d2f4716f71bae7a55a9f4153/src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php#L52

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php (hunk 1 above)
Comment (Tobion) [reply src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php]: Also what does a node without a name do?

Comment (ro0NL) [reply src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php]: Ok my bad.. it's clarified only on 4.0: https://github.com/symfony/symfony/blob/4.0/src/Symfony/Component/Config/Definition/BaseNode.php#L39

Should be backported to 2.7 as `string|null` AFAIK, this being the first instance :)

In general it's used to create a prototyped node: https://github.com/symfony/symfony/blob/4.0/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L77

I think it makes sense to keep string API when using a named factory: `scalarNode('string')` vs `node(null, 'scalar')`. Which seems to be the case already; https://github.com/symfony/symfony/blob/4.0/src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php#L52

**Comment by ro0NL** [comment]
@Tobion im leaning to revert the last `string|null` fix on node names for now. It's a dirty road.

Still bugged on master:

https://github.com/symfony/symfony/blob/cd5f4105a43208632c8b62fd75b4dad53dcd8a41/src/Symfony/Component/Config/Definition/BaseNode.php#L42-L45

^ strpos on nullable

https://github.com/symfony/symfony/blob/cd5f4105a43208632c8b62fd75b4dad53dcd8a41/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L381-L383

^ not sure.. maybe dragons, maybe not.

Put different im not sure about fix that on 2.7... but backporting `string|null` sounds reasonable.

**Comment by ro0NL** [comment]
Status: needs review

to go along with #26335, applies until 3.4

**Comment by nicolas-grekas** [comment]
@Tobion OK for you?

**Review by Tobion — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @ro0NL.

**Pull request merged**